tests/harness/utils.py

```
@@ -73,6 +73,30 @@ def get_server_metrics_url(self, *server_types: str) -> list[str]:
 
+    def recorded_headers(self) -> list[dict[str, Any]]:
+        """Return the inbound headers the mock server saw on inference paths.
+
+        Each entry is `{"path": "...", "headers": {<lowercased name>: <value>}}`.
+        Useful for asserting end-to-end header fidelity through the aiohttp
+        transport. Call `clear_recorded_headers()` first to scope the result
+        to a single test (the mock server is shared across the package).
+        """
+        import urllib.request
+
+        with urllib.request.urlopen(f"{self.url}/test/recorded_headers") as resp:
```

Ruff (0.15.14)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)